File: app/src/main/java/codes/wokstym/cookingrecipes/views/RecipeListActivity.java
package codes.wokstym.cookingrecipes.views;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.recyclerview.widget.LinearLayoutManager;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import codes.wokstym.cookingrecipes.R;
import codes.wokstym.cookingrecipes.components.RecipeAdapter;
import codes.wokstym.cookingrecipes.components.VerticalSpaceItemDecoration;
import codes.wokstym.cookingrecipes.databinding.ActivityRecipeListBinding;
import codes.wokstym.cookingrecipes.models.RecipeDto;
import codes.wokstym.cookingrecipes.models.ShoppingListDto;
import codes.wokstym.cookingrecipes.tasks.GetRecipesTask;
import codes.wokstym.cookingrecipes.tasks.GetShoppingListTask;
import codes.wokstym.cookingrecipes.utils.ComponentsUtils;

import static codes.wokstym.cookingrecipes.utils.ConversionUtils.dpToPx;
import static codes.wokstym.cookingrecipes.utils.ExtraUtils.RECIPE_EXTRA;
import static codes.wokstym.cookingrecipes.utils.ExtraUtils.SHOPPING_LIST_EXTRA;

public class RecipeListActivity extends AppCompatActivity {

    private RecipeAdapter recipeAdapter;
    private ActivityRecipeListBinding binding;
    List<RecipeDto> recipeList;
    ActionMode actionMode;
    ActionCallback actionCallback;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityRecipeListBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        actionCallback = new ActionCallback();

        initSupportBar();
        initListeners();

        fetchRecipes();
    }

    private void initSupportBar() {
        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setTitle(getResources().getString(R.string.recipe_list));
    }

    private void initListeners() {
        binding.recipeListButton.setOnClickListener(v -> fetchShoppingList());
    }

    private void fetchRecipes() {
        showProgress();
        new GetRecipesTask(this).execute();
    }

    public void showRecipes(List<RecipeDto> recipeList) {
        this.recipeList = recipeList;
        sortRecipeList(recipeList);

        prepareAdapter(recipeList);
        prepareRecipeRecyclerView();
    }

    private void sortRecipeList(List<RecipeDto> changesList) {
        changesList.sort((o1, o2) -> o1.getMealTime().getOrder() - o2.getMealTime().getOrder());
    }

    private void prepareAdapter(List<RecipeDto> recipeList) {
        recipeAdapter = new RecipeAdapter(this, recipeList);
        recipeAdapter.setOnItemClick(new RecipeAdapter.OnItemClick() {

            @Override
            public void onItemClick(View view, RecipeDto recipe, int position) {
                if (recipeAdapter.selectedItemCount() > 0) {
                    toggleActionBar(position);
                } else {
                    goToRecipeDetails(position, recipeList);
                }
            }

            @Override
            public void onLongPress(View view, RecipeDto recipe, int position) {
                toggleActionBar(position);
            }
        });
    }

    private void prepareRecipeRecyclerView() {
        binding.recipeRecyclerView.setAdapter(recipeAdapter);
        binding.recipeRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        VerticalSpaceItemDecoration dividerItemDecoration = new VerticalSpaceItemDecoration(dpToPx(5));
        binding.recipeRecyclerView.addItemDecoration(dividerItemDecoration);
    }

    private void toggleActionBar(int position) {
        if (actionMode == null) {
            actionMode = startSupportActionMode(actionCallback);
        }
        toggleSelection(position);
    }

    private void goToRecipeDetails(int position, List<RecipeDto> recipeList) {
        Intent intent = new Intent(this, RecipeDetails.class);
        intent.putExtra(RECIPE_EXTRA, recipeList.get(position));
        startActivity(intent);
    }

    private void toggleSelection(int position) {
        recipeAdapter.toggleSelection(position);
        int count = recipeAdapter.selectedItemCount();
        if (count == 0) {
            actionMode.finish();
        } else {
            actionMode.setTitle(String.valueOf(count));
            actionMode.invalidate();
        }
    }

    private void fetchShoppingList() {
        List<RecipeDto> selectedItemPositions = recipeAdapter.getSelectedItems();
        List<UUID> recipesUUIDs = selectedItemPositions.stream()
                .map(RecipeDto::getId)
                .collect(Collectors.toList());

        showProgress();
        new GetShoppingListTask(this).execute(recipesUUIDs);
    }

    public void hideProgress() {
        binding.progressBar.setVisibility(View.GONE);
    }

    private void showProgress() {
        binding.progressBar.setVisibility(View.VISIBLE);
    }

    public void showShoppingList(ShoppingListDto shoppingList) {
        Intent intent = new Intent(this, ShoppingListActivity.class);
        intent.putExtra(SHOPPING_LIST_EXTRA, shoppingList);
        startActivity(intent);
    }

    private class ActionCallback implements ActionMode.Callback {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            ComponentsUtils.toggleStatusBarColor(RecipeListActivity.this, R.color.DarkSlateGray);
            mode.getMenuInflater().inflate(R.menu.menu, menu);
            binding.recipeListButton.setVisibility(View.VISIBLE);
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            recipeAdapter.clearSelection();
            actionMode = null;
            binding.recipeListButton.setVisibility(View.GONE);
            ComponentsUtils.toggleStatusBarColor(RecipeListActivity.this, R.color.colorPrimary);
        }
    }

}

File: app/src/main/java/codes/wokstym/cookingrecipes/views/RecipeListActivity.kt
package codes.wokstym.cookingrecipes.views

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import codes.wokstym.cookingrecipes.R
import codes.wokstym.cookingrecipes.components.RecipeAdapter
import codes.wokstym.cookingrecipes.databinding.ActivityRecipeListBinding
import codes.wokstym.cookingrecipes.models.RecipeDto
import codes.wokstym.cookingrecipes.models.ShoppingListDto
import codes.wokstym.cookingrecipes.tasks.GetRecipesTask
import codes.wokstym.cookingrecipes.tasks.GetShoppingListTask
import codes.wokstym.cookingrecipes.utils.*
import com.squareup.picasso.Picasso
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class RecipeListActivity : AppCompatActivity() {

    private lateinit var binding: ActivityRecipeListBinding
    private var actionCallback: ActionCallback = ActionCallback()

    private lateinit var recipeAdapter: RecipeAdapter
    private var actionMode: ActionMode? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityRecipeListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        fun initSupportBar() {
            setSupportActionBar(binding.toolbar)
            supportActionBar?.setDisplayShowTitleEnabled(false)
        }

        fun initListeners() {
            binding.recipeListButton.setOnClickListener { fetchShoppingList() }
        }

        fun setProfile() { // TODO: replace with user data
            @SuppressLint("SetTextI18n")
            binding.userName.text = "Grzegorz Poręba"
            Picasso.with(this)
                .load("https://avatars.githubusercontent.com/u/44115112?s=460&u=2fea6d808fb949060aa499dad3e3365608bb5c40&v=4")
                .into(binding.userProfile)
        }

        initSupportBar()
        setProfile()
        initListeners()
        fetchRecipes()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) =
        when (item.itemId) {
            R.id.add -> {
                startAddRecipeActivity()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }

    fun startShoppingList(shoppingList: ShoppingListDto) {
        startShoppingListActivity(shoppingList, ArrayList(recipeAdapter.selectedItems))
    }

    private fun fetchRecipes() {
        showProgressBar()
        GetRecipesTask(this).execute()
    }

    private fun fetchShoppingList() {
        val recipesUUIDs = recipeAdapter.selectedItems
            .map(RecipeDto::id)
            .toList()
        showProgressBar()
        GetShoppingListTask(this).execute(recipesUUIDs)
    }

    fun showRecipes(recipeList: List<RecipeDto>) {
        val sorted = recipeList.sorted()
        prepareAdapter(sorted)
        prepareRecyclerView()
    }

    private fun prepareAdapter(recipeList: List<RecipeDto>) {
        recipeAdapter = RecipeAdapter(this@RecipeListActivity, recipeList, object : RecipeAdapter.OnItemClick {
            override fun onItemClick(view: View, recipe: RecipeDto, position: Int) {
                if (recipeAdapter.selectedItemsCount > 0) {
                    toggleActionBar(position)
                } else {
                    startRecipeDetailsActivity(recipe)
                }
            }

            override fun onLongPress(view: View, recipe: RecipeDto, position: Int) {
                toggleActionBar(position)
            }
        })
    }

    private fun prepareRecyclerView() =
        binding.recipeRecyclerView.apply {
            adapter = recipeAdapter
            layoutManager = GridLayoutManager(this@RecipeListActivity, 2)
            (itemAnimator as DefaultItemAnimator).supportsChangeAnimations = false
            addGridSpaceDivider(10.0)
        }


    private fun toggleActionBar(position: Int) {
        if (actionMode == null) {
            actionMode = startSupportActionMode(actionCallback)
        }
        toggleSelection(position)
    }

    private fun toggleSelection(position: Int) {
        recipeAdapter.toggleSelection(position)
        val count = recipeAdapter.selectedItemsCount
        if (count == 0) {
            actionMode?.finish()
        } else {
            actionMode?.title = count.toString()
            actionMode?.invalidate()
        }
    }

    inner class ActionCallback : ActionMode.Callback {

        override fun onCreateActionMode(mode: ActionMode, menu: Menu?): Boolean {
            binding.recipeListButton.show()
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?) = false

        override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?) = false

        override fun onDestroyActionMode(mode: ActionMode?) {
            recipeAdapter.clearSelection()
            actionMode = null
            binding.recipeListButton.hide()
        }
    }
}